pyfarm/agent/service.py

```diff
@@ -472,151 +483,183 @@ def start(self, shutdown_events=True, http_server=True):
                 self.callback_agent_id_set, shutdown_events=shutdown_events))
         return self.post_agent_to_master()
 
+    @inlineCallbacks
     def stop(self):
         """
         Internal code which stops the agent.  This will terminate any running
         processes, inform the master of the terminated tasks, update the
         state of the agent on the master.
         """
+        yield self.stop_lock.acquire()
+        if self.stopped:
+            yield self.stop_lock.release()
+            svclog.warning("Agent is already stopped")
+            returnValue(None)
+
         svclog.info("Stopping the agent")
 
-        # If this is the first time we're calling stop() then
-        # setup a function call to remove the pidfile when the
-        # process exits.
-        if not self.shutting_down:
-            self.shutting_down = True
-            self.shutdown_timeout = (
-                datetime.utcnow() + timedelta(
-                    seconds=config["agent_shutdown_timeout"]))
+        self.shutting_down = True
+        self.shutdown_timeout = (
+            datetime.utcnow() + timedelta(
+                seconds=config["agent_shutdown_timeout"]))
 
-            utility.remove_file(
-                config["agent_lock_file"], retry_on_exit=True, raise_=False)
+        utility.remove_file(
+            config["agent_lock_file"], retry_on_exit=True, raise_=False)
 
-            svclog.debug("Stopping execution of jobtypes")
-            for uuid, jobtype in config["jobtypes"].items():
+        svclog.debug("Stopping execution of jobtypes")
+        for jobtype_id, jobtype in config["jobtypes"].items():
+            try:
                 jobtype.stop()
+            except Exception as error:  # pragma: no cover
+                svclog.warning(
+                    "Error while calling stop() on %s (id: %s): %s",
+                    jobtype, jobtype_id, error
+                )
+                config["jobtypes"].pop(jobtype_id)
 
-            def wait_on_stopped():
-                svclog.info("Waiting on %s job types to terminate",
-                            len(config["jobtypes"]))
-
-                for jobtype_id, jobtype in config["jobtypes"].copy().items():
-                    if not jobtype._has_running_processes():
-                        svclog.error(
-                            "%r has not removed itself, forcing removal",
-                            jobtype)
-                        config["jobtypes"].pop(jobtype_id)
-
-                if config["jobtypes"]:
-                    reactor.callLater(1, wait_on_stopped)
-                elif self.agent_api() is not None:
-                    self.post_shutdown_to_master()
-                else:
-                    reactor.stop()
+        svclog.info(
+            "Waiting on %s job types to terminate", len(config["jobtypes"]))
+
+        while True:
+            # No jobtypes remain
+            if not config["jobtypes"]:
+                break
```

Comment:
Can't you rewrite the while loop instead as
```while config["jobtypes"] and datetime.utcnow() <= self.shutdown_timeout```?

A ```while True``` in a place that's not actually supposed to be an endless loop seems kinda wrong to me...